dev_docs/kibana_platform_plugin_intro.mdx

```diff
@@ -11,45 +11,35 @@ From an end user perspective, Kibana is a tool for interacting with Elasticsearc
 
 ![Kibana personas](assets/kibana_platform_plugin_end_user.png)
+ 
+## Platform services
 
-## Plugins vs The Platform
+Plugins have access to three kinds of public services:
 
-The core platform provides the most basic and fundamental tools neccessary for building a plugin, like creating saved objects,
-routing, application registration, and notifications. The Core platform is not a plugin itself, although
-there are some plugins that provide platform functionality. For example, the 
-<DocLink id="kibDataPlugin" text="data plugin"/> provides basic utilities to search, query, and filter data in Elasticsearch. 
-This code is not part of Core, but is still fundamental for building a plugin,
- and we strongly encourage using this service over querying Elasticsearch directly. 
-
-<DocCallOut title="Three kinds of public services">
-We currently have three kinds of public services:
-
- - platform services provided by `core`
- - platform services provided by plugins, that can, and should, be used by every plugin (e.g. <DocLink id="kibDataPlugin" text="data"/>) .
- - shared services provided by plugins, that are only relevant for only a few, specific plugins (e.g. "presentation utils"). 
-
-Two common questions we encounter are:
+ - Platform services provided by `core` (<DocLink id="kibPlatformIntro" section="core-services" text="Core services"/>)
+ - Platform services provided by plugins (<DocLink id="kibPlatformIntro" section="platform-plugins" text="Platform plugins"/>)
+ - Shared services provided by plugins, that are only relevant for only a few, specific plugins (e.g. "presentation utils"). 
 
-1. Which services are platform services? 
-2. What is the difference between platform code supplied by core, and platform code supplied by plugins?
+ The first two items are what make up "Platform services".
 
-We don't have great answers to those questions today. Currently, the best answers we have are:
+<DocCallOut title="What is the difference between services provided by platform plugins, and those by core?">
 
-1. Platform plugins are _usually_ plugins that are managed by the Platform Group, but we are starting to see some exceptions.
-2. `core` code contains the most fundamental and stable services needed for plugin development. Everything else goes in a plugin.
+We try to put only the most stable and fundamental code into `Core`, while more application focused functionality goes in a plugin, but the heuristic isn't
+clear, and we haven't done a great job of sticking to it. For example, notifications and toasts are core services, but data and search are plugin services.
 
-We will continue to focus on adding clarity around these types of services and what developers can expect from each.
-
-</DocCallOut>
+Today it looks like the left, where we should probably be closer to the diagram on the right. Core should be an implementation detail of the platform, and plugins 
+shouldn't need to understand the difference.
```

Comment:
**disclaimer:** I'm going to say a lot obvious things that we all know, but I find it helpful to bring that context in to this discussion.

> From the change in the PR:
> > Core should be an implementation detail of the platform, and plugins shouldn't need to understand the difference.
>
> @rudolf's comment:
> > There is a strong reason for the separation between core and platform that's not just conway's law in action. They have different API stability goals.

I really like the idea of using "API stability goals" as a primary lens for deciding how Platform code is organized and exposed to applications. However, I don't think we've historically described the Core vs. plugins organization in this way before, though it does closely mirror how we've gone about making individual decisions in practice.

Stacey's changes here reflect some of the discussions we've been having around evolving the Platform & Developer Experience of consuming the Platform to help applications teams locate useful APIs, abstractions, and re-usable components, but also signalling maturity of these APIs in order to avoid creating new undue or unexpected maintenance.

I think there is more than one issue being conflated here and though they all overlap a bit, I'd like to tease them apart a bit to dissect each issue separately:
1) Code organization and how this intersects with Conway's Law
2) API stability and maintenance, but also API flexibility and escape hatches; two sides of the same coin.
3) Discoverability of shared services by application teams

### Code organization and Conway's Law

Assuming Conway's Law is true, the API surfaces between components should  (and inevitably **will**) reflect the team interfaces/APIs between teams. Creating clear interfaces & abstractions in code between teams' components to communicate with one another allows us to reduce communication overhead IRL. Preserving clear boundaries between different domains in code has many benefits, and was a clear goal and outcome of the New Platform project.

As we evolve the Platform, I think it'd be very prudent of us to maintain these boundaries, else we risk spaghetti code proliferation all over again.

### API stability & flexibility

Creating shared services comes with a different set of challenges than application development. Balancing long-term maintenance with velocity of change becomes a more difficult dilemma. Thus far, our primary strategy on the Core team has been to keep the `src/core` codebase as thin as possible and provide extension points for other systems. This strategy is also known as the ["Thinnest Viable Platform"](https://teamtopologies.com/key-concepts-content/what-is-a-thinnest-viable-platform-tvp) and there are many benefits. It allows us to create a smaller, more stable API, but this can only be achieved by maximizing flexibility so that changes to the platform are not needed frequently in order to support application development.

Where we're at now is the "fractal platform" stage where we need to provide both lower-level, flexible APIs along with higher-level, APIs with rails that encourage developers to follow best practices and write less code to achieve what the need.

I think the problem we're facing now is that our higher-level APIs are becoming much more mature, yet are not widely used enough. It seems to me that if we want to encourage developers to start using the higher-level APIs, we should start to change where stability is most important to the higher-level services. This would actually allow us to more easily evolve the lower-level Core APIs by moving much of the maintenance burden to the higher-level services and those services being the primary consumers of the lower-level Core APIs.

We'll still need the escape hatch of Core APIs for when application teams need to do something that a higher-level service doesn't provide, but ideally the number of applications integrating directly with Core for more complex use cases becomes much smaller.

### Discoverability of shared services

With many more mature services emerging, we should encourage adoption by exposing these services by default. We're at the point where disabling any of these plugins doesn't really make any sense anymore, so does it make sense to still require applications to declare dependencies on these services? To me, that seems like an additional hurdle that hampers discoverability, however it does improve our ability to see the dependency graph between applications and the plugin services they depend on.

## Putting it all together

It seems to me that we can achieve both a directory and dependency hierarchy that reflects the organization we want while also providing improved discoverability of shared services and clear stability signals. I still strongly believe the [organize by domain](https://github.com/elastic/kibana/issues/71566) structure will benefit us a lot in this regard and the recent licensing changes will give us more flexibility in how we implement this. This will also help us think about the interface between Core and higher-level services. 

All this to say that I think there are iterative steps that we have to take to improve discoverability and stability simultaneously. Starting with documentation of 'platform plugin services' (with `alpha`, `beta`, `stable` tags?) makes total sense to me, and I think it's useful to start pushing the concept of these services as being part of a larger Platform that shares similar API designs, idioms, and concepts.

As we evolve we can start to consider new ways of exposing the larger platform to developers, along with signaling the stability of an API in code. I don't think we have to "promote" these things to Core in order to achieve that, and in fact, I am strongly opposed to that due to the Conway's law issues mentioned above. We need to maintain both clear boundaries _and_ discoverability.

Reply:
What @joshdover said :)

Reply:
++ to @joshdover

> So what does it mean for developers? They probably just need to know the following things:
> 1. How stable is this API (we probably need a roadmap for promoting core to stable)

Agreed! But also agree with @joshdover that I think a more future proof approach will be via @ tags. Increasing the stability of an API shouldn't require a breaking change or large refactor (and moving a plugin into core would).  Using tags will also allow us to mix new APIs alongside stable APIs, within the same plugin, which will let us keep our "organize by domain" principle as we try out new features (for example, async search).

I agree though we can take this discussion offline, and have removed it from the doc.